tests/TestCase/src/Controller/BlogPostsControllerTest.php

```diff
@@ -29,41 +32,43 @@ public function setUp(): void
 
     public function testBlogPostDetailOnlinePublicLoggedOut()
     {
-        $this->httpClient->get($this->Slug->getBlogPostDetail(2, 'Demo Blog Artikel'));
-        $this->assert200OkHeader();
+        $_SERVER['REQUEST_URI'] = $this->Slug->getBlogPostDetail(2, 'Demo Blog Artikel');
+        $this->get($_SERVER['REQUEST_URI']);
+        $this->assertResponseCode(200);
     }
 
     public function testBlogPostDetailOfflinePublicLoggedOut()
     {
         $blogPostId = 2;
         $this->changeBlogPost($blogPostId, 0, 0, 0);
-        $this->httpClient->get($this->Slug->getBlogPostDetail($blogPostId, 'Demo Blog Artikel'));
-        $this->assert404NotFoundHeader();
+        $this->get($this->Slug->getBlogPostDetail($blogPostId, 'Demo Blog Artikel'));
+        $this->assertResponseCode(404);
     }
 
     public function testBlogPostDetailOnlinePrivateLoggedOut()
     {
         $blogPostId = 2;
         $this->changeBlogPost($blogPostId, 1);
-        $this->httpClient->followOneRedirectForNextRequest();
-        $this->httpClient->get($this->Slug->getBlogPostDetail($blogPostId, 'Demo Blog Artikel'));
-        $this->assertAccessDeniedWithRedirectToLoginForm();
+        $this->get($this->Slug->getBlogPostDetail($blogPostId, 'Demo Blog Artikel'));
+        $this->Slug->getLogin($blogPostId, 'Demo Blog Artikel');
```

Comment:
You need to assert the result of Slug->getLogin()
eg. `$this->assertRedirectContains($this->Slug->getLogin($blogPostId, 'Demo Blog Artikel'));`
eventually you nedd to move urlencode from here to getLogin() https://github.com/foodcoopshop/foodcoopshop/blob/81bd838dd1071efeb54b9756583302ba06d0be07/templates/Customers/login.php#L32 into getLogin()

Reply:
It should be something like `$this->assertRedirectContains($this->Slug->getLogin('%2Faktuelles%2F'.$blogPostId.'-Demo-Blog-Artikel'));`. What do you think?

Reply:
The result should be the same as this:
`'/anmelden?redirect=%2Faktuelles%2F2-Demo-Blog-Artikel'`
And please also use SlugHelper->getBlogPostDetail and do not manually build the url.